README.md
# immaculater_dart
Dart-language code encapsulating
[Immaculater's](https://github.com/chandler37/immaculater) mergeprotobufs API
for reading and writing personal to-do lists with actions, contexts, projects,
folders, and notes.

One future direction is to use this library in a new Immaculater Flutter
app. Instead of dealing with little bits of data, you'll be sending back and
forth the entire to-do list since they're typically small enough for that to be
performant in the asynchronous world of Flutter (it is just for one person,
after all).

Users can then have their choice of the
[Immaculater](https://github.com/chandler37/immaculater) classic web
application or a Flutter mobile/desktop/web app. They will be able to
simultaneously use a Flutter iOS app and a Flutter desktop app thanks to
mergeprotobufs' ability to synchronize various clients.

To understand this library you must understand the mergeprotobufs API which is
documented in the
[source](https://github.com/chandler37/immaculater/blob/master/todo/views.py)
(search for `def mergeprotobufs(request):`).

In short, though, mergeprotobufs is the way to read and to write. A write is
also a read, sometimes, if another device has changed the to-do list in the
interim.


## Is it any good?

More testing is necessary before data loss is unlikely.


## Staying up to date with Immaculater's protocol

[Immaculater](https://github.com/chandler37/immaculater) generates Dart
protocol buffers as a matter of course. Just copy them here with
`pyatdl.proto`.


## What to type

If you're using this library, watch this space for
instructions. TODO(chandler37): Give instructions beyond "Call merge()".

If you're developing this library, type this:

`make lint fmt test`

Which will only work if you have the Dart SDK installed. (I used the Dart SDK
via [Homebrew](https://brew.sh/), not the Dart inside of Flutter's SDK.)


## How does it work under the hood?

We use a mix of JSON (for errors returned), HTTP 204 NO CONTENT, and [protocol
buffers](https://developers.google.com/protocol-buffers/docs/darttutorial) for
the input and output of mergeprotobufs. It is not gRPC.

To make progress, learn how `immaculater_dart_test.dart` works. There is a
skipped test that can be used to make the test suite hit a live backend. Record
that interaction in a "cassette". The tests should not require network access.


## Credit where credit is due

Created from templates made available by Stagehand under a BSD-style
[license](https://github.com/dart-lang/stagehand/blob/master/LICENSE).


CHANGELOG.md
## 0.7.3

        - Updated pyatdl.proto to match the generated code

## 0.7.2

        - More modern generated protocol buffer Dart files, which is hopefully a noop or a bug fix.

## 0.7.1

        - Bug fix for HTTP 204 response from mergeprotobufs

## 0.7.0

        - UnauthenticatedException is also used for creating JWTs.

## 0.6.0

        - UnauthenticatedException is a new subclass of ApiException for
        HTTP 403 responses (which confuse authorization with
        authentication (see HTTP 401) but really are about
        authentication).

## 0.5.0

        - Protobufs now feature a message UserInfoAndToDoList.

## 0.4.0

        - JSON Web Token (JWT) authentication is supported.

## 0.3.0

        - Merely updates the protocol with two new booleans for
        MergeToDoListRequest.

## 0.2.0

        - Downgrades meta dependency to play nice with Flutter. Many
        incremental improvements and new APIs. Many new tests.

## 0.1.0

        - Initial version (thank you to the Stagehand code generator),
        shaped with care to perhaps work for a 'we never need to merge'
        use case. Perhaps even when a merge is required.


test/cassettes/read_before_write0.dart
// This HTTP response body was created by sending in emptyMergeRequest when
// there was a starter template in the database.

// I hand-verified that this, 015bf1f05eb048f4afbaa2cc1862bba667b0c51a, matches the uncompressed pyatdl.ToDoList
var sha1Checksum = "015bf1f05eb048f4afbaa2cc1862bba667b0c51a";

var b64 =
    r"CigwMTViZjFmMDVlYjA0OGY0YWZiYWEyY2MxODYyYmJhNjY3YjBjNTFhEoEJCjIKLAgAEh0Ig+bnoprN5QIQ////////////ARj25ueims3lAhoHCgVpbmJveCABEAAYARLYBAonCAASHQiK5+eims3lAhD///////////8BGMCD6KKazeUCGgIKACACGkMKPQgAEh0I7//noprN5QIQ////////////ARjXgOiims3lAhoPCg1taXNjZWxsYW5lb3VzIMqll6fQ8pjS2gEQABgBGucDClAIABIdCLyC6KKazeUCEP///////////wEY7Y7ooprN5QIaIgogbGVhcm4gaG93IHRvIHVzZSB0aGlzIHRvLWRvIGxpc3Qgo7P3nZayk8epARAAGAEiewp3CAASHQiYhuiims3lAhD///////////8BGOKG6KKazeUCGkkKR1dhdGNoIHRoZSB2aWRlbyBvbiB0aGUgIkhlbHAiIHBhZ2UgLS0gZmluZCBpdCBvbiB0aGUgdG9wIG5hdmlnYXRpb24gYmFyINfaoarLneberAEYACJlCmEIABIdCOSJ6KKazeUCEP///////////wEYrorooprN5QIaNAoyUmVhZCB0aGUgYm9vayAiR2V0dGluZyBUaGluZ3MgRG9uZSIgYnkgRGF2aWQgQWxsZW4grJaZ0dGs9b5qGAAiqgEKpQEIABIdCI+O6KKazeUCEP///////////wEY5Y7ooprN5QIaeAp2QWZ0ZXIgcmVhZGluZyB0aGUgYm9vaywgdHJ5IG91dCBhIFdlZWtseSBSZXZpZXcgLS0gb24gdGhlIHRvcCBuYXZpZ2F0aW9uIGJhciwgZmluZCBpdCB1bmRlcm5lYXRoIHRoZSAiT3RoZXIiIGRyb3AtZG93biCwqqm0sYu3r0kYABrvAwo4CAASHQj15+eims3lAhD///////////8BGL7856KazeUCGgoKCENvbnRleHRzIKXhhqCY1/TK5gESPQo5CAASHQjR7eeims3lAhD///////////8BGJfu56KazeUCGgsKCUBjb21wdXRlciCuhtfU2taQi60BEAESOgo2CAASHQih8Oeims3lAhD///////////8BGOfw56KazeUCGggKBkBwaG9uZSCszY/Fp5XhgOsBEAESOAo0CAASHQik8ueims3lAhD///////////8BGOjy56KazeUCGgcKBUBob21lIKny2KaDqqKLDxABEjkKNQgAEh0IkPXnoprN5QIQ////////////ARjc9eeims3lAhoHCgVAd29yayDMndCEqt65lLYBEAESPgo6CAASHQjf9+eims3lAhD///////////8BGKL456KazeUCGgwKCkB0aGUgc3RvcmUg/4TS0em4t6+VARABEkEKPQgAEh0I1/nnoprN5QIQ////////////ARi7/eeims3lAhoQCg5Ac29tZWRheS9tYXliZSCJhK+52M/ovT4QABJACjwIABIdCMf756KazeUCEP///////////wEYn/7noprN5QIaDgoMQHdhaXRpbmcgZm9yIMru8PejoJOj7gEQABgBed7A/srO+u3+";

var textOfMergeToDoListResponse = '''
sha1Checksum: $sha1Checksum
toDoList: {
  inbox: {
    common: {
      isDeleted: false
      timestamp: {
        ctime: 1572755356644099
        dtime: -1
        mtime: 1572755356644214
      }
      metadata: {
        name: inbox
      }
      uid: 1
    }
    isComplete: false
    isActive: true
  }
  root: {
    common: {
      isDeleted: false
      timestamp: {
        ctime: 1572755356644234
        dtime: -1
        mtime: 1572755356647872
      }
      metadata: {
        name:
      }
      uid: 2
    }
    projects: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356647407
          dtime: -1
          mtime: 1572755356647511
        }
        metadata: {
          name: miscellaneous
        }
        uid: -2691917185577266486
      }
      isComplete: false
      isActive: true
    }
    projects: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356647740
          dtime: -1
          mtime: 1572755356649325
        }
        metadata: {
          name: learn how to use this to-do list
        }
        uid: -6228955947767834205
      }
      isComplete: false
      isActive: true
      actions: {
        common: {
          isDeleted: false
          timestamp: {
            ctime: 1572755356648216
            dtime: -1
            mtime: 1572755356648290
          }
          metadata: {
            name: Watch the video on the "Help" page -- find it on the top navigation bar
          }
          uid: -5999470986166506153
        }
        isComplete: false
      }
      actions: {
        common: {
          isDeleted: false
          timestamp: {
            ctime: 1572755356648676
            dtime: -1
            mtime: 1572755356648750
          }
          metadata: {
            name: Read the book "Getting Things Done" by David Allen
          }
          uid: 7673523970316323628
        }
        isComplete: false
      }
      actions: {
        common: {
          isDeleted: false
          timestamp: {
            ctime: 1572755356649231
            dtime: -1
            mtime: 1572755356649317
          }
          metadata: {
            name: After reading the book, try out a Weekly Review -- on the top navigation bar, find it underneath the "Other" drop-down
          }
          uid: 5286905296357840176
        }
        isComplete: false
      }
    }
  }
  ctxList: {
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356645073
          dtime: -1
          mtime: 1572755356645143
        }
        metadata: {
          name: @computer
        }
        uid: -5974514507641928914
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356645409
          dtime: -1
          mtime: 1572755356645479
        }
        metadata: {
          name: @phone
        }
        uid: -1512782132116461908
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356645668
          dtime: -1
          mtime: 1572755356645736
        }
        metadata: {
          name: @home
        }
        uid: 1087207337633397033
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356646032
          dtime: -1
          mtime: 1572755356646108
        }
        metadata: {
          name: @work
        }
        uid: -5320749029987709236
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356646367
          dtime: -1
          mtime: 1572755356646434
        }
        metadata: {
          name: @the store
        }
        uid: -7683460069187091841
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356646615
          dtime: -1
          mtime: 1572755356647099
        }
        metadata: {
          name: @someday/maybe
        }
        uid: 4502370912509346313
      }
      isActive: false
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572755356646855
          dtime: -1
          mtime: 1572755356647199
        }
        metadata: {
          name: @waiting for
        }
        uid: -1277248772270966966
      }
      isActive: false
    }
1: [8, 0, 18, 29, 8, 245, 231, 231, 162, 154, 205, 229, 2, 16, 255, 255, 255, 255, 255, 255, 255, 255, 255, 1, 24, 190, 252, 231, 162, 154, 205, 229, 2, 26, 10, 10, 8, 67, 111, 110, 116, 101, 120, 116, 115, 32, 165, 225, 134, 160, 152, 215, 244, 202, 230, 1]
  }
}
starterTemplate: true
sanityCheck: -77129852519530274
'''
    .replaceAll('name:\n', 'name: \n'); // EditorConfig detests trailing whitespace


test/cassettes/read_before_write1.dart
// This HTTP response body was created by sending in emptyMergeRequest when
// there was a starter template in the database.

var sha1Checksum = "542edb871f9d4c161999e881a7846e28bd12e4cc";

var b64 =
    r"Cig1NDJlZGI4NzFmOWQ0YzE2MTk5OWU4ODFhNzg0NmUyOGJkMTJlNGNjEv0ICjIKLAgAEh0IxNjLw87Q5QIQ////////////ARiJ2svDztDlAhoHCgVpbmJveCABEAAYARLWBAonCAASHQie2svDztDlAhD///////////8BGNiCzMPO0OUCGgIKACACGkIKPAgAEh0IzvzLw87Q5QIQ////////////ARiX/svDztDlAhoPCg1taXNjZWxsYW5lb3VzILjR4obliYDnSBAAGAEa5gMKUAgAEh0IloHMw87Q5QIQ////////////ARidlszDztDlAhoiCiBsZWFybiBob3cgdG8gdXNlIHRoaXMgdG8tZG8gbGlzdCCx9cGC5b3uuZkBEAAYASJ6CnYIABIdCOSIzMPO0OUCEP///////////wEY8InMw87Q5QIaSQpHV2F0Y2ggdGhlIHZpZGVvIG9uIHRoZSAiSGVscCIgcGFnZSAtLSBmaW5kIGl0IG9uIHRoZSB0b3AgbmF2aWdhdGlvbiBiYXIg1cuBjOKBvLFEGAAiZQphCAASHQj5jszDztDlAhD///////////8BGOePzMPO0OUCGjQKMlJlYWQgdGhlIGJvb2sgIkdldHRpbmcgVGhpbmdzIERvbmUiIGJ5IERhdmlkIEFsbGVuILn7h4W2pMH7IhgAIqoBCqUBCAASHQiHlczDztDlAhD///////////8BGI2WzMPO0OUCGngKdkFmdGVyIHJlYWRpbmcgdGhlIGJvb2ssIHRyeSBvdXQgYSBXZWVrbHkgUmV2aWV3IC0tIG9uIHRoZSB0b3AgbmF2aWdhdGlvbiBiYXIsIGZpbmQgaXQgdW5kZXJuZWF0aCB0aGUgIk90aGVyIiBkcm9wLWRvd24gnJrB2P61hagrGAAa7QMKNwgAEh0IgNvLw87Q5QIQ////////////ARjX98vDztDlAhoKCghDb250ZXh0cyDGnPvPjcv9vw4SPQo5CAASHQj14svDztDlAhD///////////8BGOTjy8PO0OUCGgsKCUBjb21wdXRlciDdod+ujOeOhPUBEAESOgo2CAASHQjh5svDztDlAhD///////////8BGMbny8PO0OUCGggKBkBwaG9uZSCSw4vEy9uo/e4BEAESOAo0CAASHQj96cvDztDlAhD///////////8BGOLqy8PO0OUCGgcKBUBob21lIKO0o/3S2eLAKBABEjkKNQgAEh0ImO3Lw87Q5QIQ////////////ARiL7svDztDlAhoHCgVAd29yayCMweazqMC88YoBEAESPQo5CAASHQi98MvDztDlAhD///////////8BGJ7xy8PO0OUCGgwKCkB0aGUgc3RvcmUg48CX4oTthYRjEAESQQo9CAASHQjK88vDztDlAhD///////////8BGJX5y8PO0OUCGhAKDkBzb21lZGF5L21heWJlIL+N74ie7azZbxAAEkAKPAgAEh0I5/bLw87Q5QIQ////////////ARiI+svDztDlAhoOCgxAd2FpdGluZyBmb3Ig5Nrb0YPIgpvvARAAed7A/srO+u3+";

var textOfMergeToDoListResponse = '''
sha1Checksum: 542edb871f9d4c161999e881a7846e28bd12e4cc
toDoList: {
  inbox: {
    common: {
      isDeleted: false
      timestamp: {
        ctime: 1572872463248452
        dtime: -1
        mtime: 1572872463248649
      }
      metadata: {
        name: inbox
      }
      uid: 1
    }
    isComplete: false
    isActive: true
  }
  root: {
    common: {
      isDeleted: false
      timestamp: {
        ctime: 1572872463248670
        dtime: -1
        mtime: 1572872463253848
      }
      metadata: {
        name:
      }
      uid: 2
    }
    projects: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463253070
          dtime: -1
          mtime: 1572872463253271
        }
        metadata: {
          name: miscellaneous
        }
        uid: 5246130952297031864
      }
      isComplete: false
      isActive: true
    }
    projects: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463253654
          dtime: -1
          mtime: 1572872463256349
        }
        metadata: {
          name: learn how to use this to-do list
        }
        uid: -7389358130384045391
      }
      isComplete: false
      isActive: true
      actions: {
        common: {
          isDeleted: false
          timestamp: {
            ctime: 1572872463254628
            dtime: -1
            mtime: 1572872463254768
          }
          metadata: {
            name: Watch the video on the "Help" page -- find it on the top navigation bar
          }
          uid: 4927764885779015125
        }
        isComplete: false
      }
      actions: {
        common: {
          isDeleted: false
          timestamp: {
            ctime: 1572872463255417
            dtime: -1
            mtime: 1572872463255527
          }
          metadata: {
            name: Read the book "Getting Things Done" by David Allen
          }
          uid: 2519488166040305081
        }
        isComplete: false
      }
      actions: {
        common: {
          isDeleted: false
          timestamp: {
            ctime: 1572872463256199
            dtime: -1
            mtime: 1572872463256333
          }
          metadata: {
            name: After reading the book, try out a Weekly Review -- on the top navigation bar, find it underneath the "Other" drop-down
          }
          uid: 3121018387074927900
        }
        isComplete: false
      }
    }
  }
  ctxList: {
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463249781
          dtime: -1
          mtime: 1572872463249892
        }
        metadata: {
          name: @computer
        }
        uid: -790316619580059427
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463250273
          dtime: -1
          mtime: 1572872463250374
        }
        metadata: {
          name: @phone
        }
        uid: -1226488879632948846
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463250685
          dtime: -1
          mtime: 1572872463250786
        }
        metadata: {
          name: @home
        }
        uid: 2918766647385315875
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463251096
          dtime: -1
          mtime: 1572872463251211
        }
        metadata: {
          name: @work
        }
        uid: -8438916659078913908
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463251517
          dtime: -1
          mtime: 1572872463251614
        }
        metadata: {
          name: @the store
        }
        uid: 7135979346292236387
      }
      isActive: true
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463251914
          dtime: -1
          mtime: 1572872463252629
        }
        metadata: {
          name: @someday/maybe
        }
        uid: 8048692751394129599
      }
      isActive: false
    }
    contexts: {
      common: {
        isDeleted: false
        timestamp: {
          ctime: 1572872463252327
          dtime: -1
          mtime: 1572872463252744
        }
        metadata: {
          name: @waiting for
        }
        uid: -1209768178931536540
      }
      isActive: false
    }
1: [8, 0, 18, 29, 8, 128, 219, 203, 195, 206, 208, 229, 2, 16, 255, 255, 255, 255, 255, 255, 255, 255, 255, 1, 24, 215, 247, 203, 195, 206, 208, 229, 2, 26, 10, 10, 8, 67, 111, 110, 116, 101, 120, 116, 115, 32, 198, 156, 251, 207, 141, 203, 253, 191, 14]
  }
}
sanityCheck: -77129852519530274
'''
    .replaceAll('name:\n', 'name: \n'); // EditorConfig detests trailing whitespace


lib/src/auth.dart
import 'dart:convert' as convert;
import 'package:http/http.dart' as http;
import 'package:meta/meta.dart';
import 'package:immaculater_dart/src/errors.dart';

abstract class Authorizer {
  void addHeaders(Map<String, String> headers);
}

class UsernamePasswordAuthorizer implements Authorizer {
  final String _username;
  final String _password;

  UsernamePasswordAuthorizer(this._username, this._password);

  @override
  void addHeaders(Map<String, String> headers) {
    String plaintextPassword = convert.base64Encode(convert.utf8.encode('$_username:$_password'));
    headers['authorization'] = 'Basic ' + plaintextPassword;
  }
}

class JsonWebTokenAuthorizer implements Authorizer {
  final String _jwt;

  JsonWebTokenAuthorizer(this._jwt);

  @override
  void addHeaders(Map<String, String> headers) {
    headers['authorization'] = 'Bearer $_jwt';
  }
}

/// Mints a new JSON Web Token (JWT) that you may later use with
/// JsonWebTokenAuthorizer. Note that the client should use
/// UsernamePasswordAuthorizer; see withClient3.
Future<String> createJsonWebToken(
    {@required String username,
    @required String password,
    @required http.Client client,
    @required String backendUrl}) async {
  var response = await client.post(backendUrl);
  if (response.statusCode == 403) {
    throw UnauthenticatedException(
        "Cannot authenticate user. A common cause is an invalid username, an administratively deactivated user, or an incorrect password.");
  }
  var json = convert.jsonDecode(response.body) as Map<String, dynamic>;
  assert(json["token"] != null);
  assert(json["token"] is String);
  String ret = json["token"] as String;
  assert(ret.isNotEmpty);
  return ret;
}


lib/immaculater_dart.dart
import 'dart:convert' as convert;
import 'dart:math' as math;
import 'dart:typed_data' show Uint8List;
import 'package:crypto/crypto.dart' as crypto;
import 'package:fixnum/fixnum.dart' show Int64;
import 'package:http/http.dart' as http;
import 'package:meta/meta.dart';
import 'package:protobuf/protobuf.dart' as protobuf;
import 'package:immaculater_dart/src/auth.dart';
import 'package:immaculater_dart/src/errors.dart';
import 'package:immaculater_dart/src/generated/core/pyatdl.pb.dart' as pb;

export 'src/auth.dart';
export 'src/errors.dart';
export 'src/generated/core/pyatdl.pb.dart';

// `library immaculater_dart` is not recommended: "When the library directive
// isn’t specified, a unique tag is generated for each library based on its path
// and filename. Therefore, we suggest that you omit the library directive from
// your code unless you plan to generate library-level documentation."
// TODO(chandler37): Look into what library-level documentation would entail.

/// A client of github.com/chandler37/immaculater which is a Django app
class DjangoClient extends http.BaseClient {
  final String userAgent;
  final Authorizer _authorizer;
  final http.Client _inner;

  DjangoClient(this._inner, this._authorizer) : userAgent = "ImmaculaterDart/0.7.3";

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) {
    request.headers['user-agent'] = userAgent;
    _authorizer.addHeaders(request.headers);
    request.headers['content-type'] = "application/x-protobuf";
    return _inner.send(request);
  }
}

/// Returns the current time suitable for pb.Timestamp usage.
Int64 now() {
  return Int64(DateTime.now().microsecondsSinceEpoch);
}

pb.Action newAction({@required String name, @required math.Random prng, String note}) {
  assert(name != null);
  var action = pb.Action();
  action.ensureCommon().ensureMetadata().name = name;
  if (note != null && note.isNotEmpty) {
    action.ensureCommon().ensureMetadata().note = note;
  }
  var ts = action.common.ensureTimestamp();
  ts.ctime = ts.mtime = now();
  action.common.uid = randomUid(prng);
  return action;
}

/// Returns a sane MergeToDoListRequest
pb.MergeToDoListRequest saneMergeRequest() {
  var req = pb.MergeToDoListRequest();
  req.sanityCheck = Int64.parseInt("18369614221190020847");
  // TODO(chandler37): we should change the sanity check value (breaking change
  // in, say, /todo/v2/mergeprotobufs) so that Int64(18369614221190020847) will
  // not be the same to make sure we're ready to run this Dart in the browser
  // where 18369614221190020847 is a double or an imprecise int because of
  // javascript numbers (2**53 limit)
  return req;
}

/// Returns the serialization of a minimal, sane MergeToDoListRequest
Uint8List emptyMergeRequest() {
  return saneMergeRequest().writeToBuffer();
}

int _next32bitInt(math.Random prng) {
  return prng.nextInt(1 << 32);
}

const inboxUid = Int64.ONE;
const rootFolderUid = Int64.TWO;
var sanityCheckForResponse = Int64.parseInt("-77129852519530274"); // no const constructor

/// Returns a uniformly pseudorandom UID u such that -2**63 <= u < 0 or 2 < u <
/// 2**63
Int64 randomUid(math.Random prng) {
  var answer = Int64.ZERO;
  while (answer == Int64.ZERO || answer == inboxUid || answer == rootFolderUid) {
    answer = Int64.fromInts(_next32bitInt(prng), _next32bitInt(prng));
  }
  return answer;
}

pb.ChecksumAndData createChecksumAndData(pb.ToDoList tdl) {
  var result = pb.ChecksumAndData();
  Uint8List bytes = tdl.writeToBuffer();
  result.payloadLength = Int64.parseInt("${bytes.length}");
  result.payload = bytes;
  result.payloadIsZlibCompressed = false;
  result.sha1Checksum = "${crypto.sha1.convert(bytes)}";
  return result;
}

/// Calls /todo/mergeprotobufs which might be a read or a write (that also does
/// a read). Returns null when and only when you did a write and the server did
/// not merge, meaning that you have the latest result and there is no need to
/// update your state with the merged pyatdl.ToDoList. Raises ApiException and
/// errors from package:http.
Future<pb.MergeToDoListResponse> merge(
    {@required String backendUrl,
    @required http.Client client,
    @required Uint8List body,
    bool verbose = false}) async {
  if (backendUrl?.isEmpty ?? true) {
    throw FormatException("Immaculater Django backend must be specified via backendUrl.");
  }
  if (verbose) {
    print('Starting HTTP request. (Do not check in tests that do actual HTTP requests.)');
  }
  var response = await client.post(backendUrl, body: body);
  if (verbose) {
    print('Response status: ${response.statusCode}');
    var pp = response.body.replaceAllMapped(RegExp(r'[^\x20-\x7E]+'), (Match m) {
      if (m.end == m.start) {
        return '\u{00BF}';
      } else {
        return '\n\u{00BF}{${m.end - m.start + 1}}';
      }
    });
    print('Response body with unprintable characters escaped: ${pp}');
  }
  if (response.statusCode == 204) {
    return null;
  }
  if (response.statusCode == 403) {
    throw UnauthenticatedException();
  }
  if (response.statusCode != 200) {
    throw ApiException(
        "unexpected httpStatusCode=${response.statusCode} with body ${response.body}");
  }
  Uint8List bodyList = Uint8List.fromList(response.body.codeUnits);
  if (verbose) {
    print('Response body as base64: ${convert.base64Encode(bodyList)}');
  }
  pb.MergeToDoListResponse resp_pb;
  try {
    resp_pb = pb.MergeToDoListResponse.fromBuffer(bodyList);
  } on protobuf.InvalidProtocolBufferException catch (e) {
    throw ApiException(
        "ill-formatted protocol buffer received; httpStatusCode=${response.statusCode}: $e");
  }
  if (resp_pb.sanityCheck != Int64.parseInt("18369614221190021342")) {
    throw ApiException("sanityCheck found was ${resp_pb.sanityCheck}");
  }
  // Or, for testing: return Future.delayed(Duration(microseconds: 2000000), () => resp_pb);
  return resp_pb;
}

/// Returns true iff respPb.sanityCheck is sane.
bool isSaneResponse(pb.MergeToDoListResponse respPb) {
  // In python3, 2**64-18369614221190021342 is 77129852519530274. Int64 is the
  // wrong abstraction for a `fixed64` field; Uint64 would be better but does
  // not exist. Integer overflow yields -77129852519530274:
  if (sanityCheckForResponse != Int64.parseInt("18369614221190021342")) {
    throw ApiException("Assertion failed re: -77129852519530274 and 18369614221190021342");
  }
  return (respPb.sanityCheck == sanityCheckForResponse);
}

/// Reads in the latest ToDoList, wrapped up in a MergeToDoListResponse.
Future<pb.MergeToDoListResponse> readToDoList(
    {@required String backendUrl, @required http.Client client}) async {
  pb.MergeToDoListResponse resp =
      await merge(backendUrl: backendUrl, client: client, body: emptyMergeRequest());
  if (resp != null && !isSaneResponse(resp)) {
    throw ApiException("Sanity check failed on response");
  }
  return resp;
}

Future<T> withClient<T>(
    Future<T> Function(
            {@required String backendUrl,
            @required http.Client client,
            @required Uint8List body,
            bool verbose})
        fn,
    {@required Authorizer authorizer,
    @required String backendUrl,
    @required Uint8List body,
    @required bool verbose}) async {
  var client = DjangoClient(http.Client(), authorizer);
  try {
    return await fn(backendUrl: backendUrl, client: client, body: body, verbose: verbose);
  } finally {
    client.close();
  }
}

Future<T> withClient2<T>(
    Future<T> Function({@required String backendUrl, @required http.Client client}) fn,
    {@required Authorizer authorizer,
    @required String backendUrl}) async {
  var client = DjangoClient(http.Client(), authorizer);
  try {
    return await fn(backendUrl: backendUrl, client: client);
  } finally {
    client.close();
  }
}

Future<T> withClient3<T>(
    Future<T> Function(
            {@required http.Client client,
            @required String backendUrl,
            @required String username,
            @required String password})
        fn,
    {@required String backendUrl,
    @required String username,
    @required String password}) async {
  var client = DjangoClient(http.Client(), UsernamePasswordAuthorizer(username, password));
  try {
    return await fn(backendUrl: backendUrl, client: client, username: username, password: password);
  } finally {
    client.close();
  }
}

// TODO(chandler37): maybe make a pull request against dart protobuf library to
// better handle int64, at least when creating ASCII prettyprint which prints
// fixed64 as negative (see sanityCheck)

// TODO(chandler37): Handle timeouts with TimeoutException. Test with
// a new /todo/timeoutplease handler perhaps.

// TODO(chandler37): Note that https://pub.dev/packages/alice exists for flutter
// inspection of http calls.
